joomla/joomla-cms
# Remove all default values for type 'text' or 'mediumtext' - redo of PR #17860

Redo of Pull Request (PR) #17860 .

_**Important for maintainers:**_ When merging, please don't do a squash commit. I don't want the commit history and so the credits for the author of the original PR being lost.

### Summary of Changes

Removed all default values for SQL type `text` or in case of MySQL also `mediumtext` for databases MySQL and PostgreSQL. For MS SQL Server nothing to be done since there `VARCHAR` columns are used in these cases.

Thanks to @eXsiLe95 for the original PR.

There were a few things missing and to be changed, see comments in the original PR.

This PR here takes over the work and completes it. Commit history is preserved so some credits still go to the original PR's author. In addition, a few inconsistencies between PostgreSQL and MySQL are fixed regarding such database columns of SQL type `text` or in case of MySQL also `mediumtext`.

### Testing Instructions

**Important:** Both of the following 2 tests should be tested on both kinds of database MySQL and PostgreSQL. If you have both kinds of database available, test both. **In any case please report back with which kind(s) of database you have tested, and in case or errors also the version(s).**

#### Test 1 - New installation

**Step 1:** Apply the patch of this PR to a clean, current staging branch or nightly build or a 3.9.15.

**Step 2:** Make a new installation.

**Step 3:** Login to backend.

**Step 4.** Go to `Extensions -> Manage -> Database`.

**Result:** No database problems shown. Database schema version is `3.9.16-2020-02-15`.

**Step 5:** Follow the testing instructions of PR #17860 , i.e.:
- Go to administrator > Modules and set it to Administrator.
- Try to create a new Administrator Menu (for example), so that there is no editor tool in the creation window (because that's where the content comes from).
- Save this module.

**Result:** No SQL error, module is saved.

**Step 6:** Verify that following works as before and there are no SQL errors shown which are related to invalid default values for columns of type `text` or in case of MySQL also `mediumtext`:
- Creating and editing categories
- Creating and editing fields and field groups
- Smart Search
- Action logs
- Content history

**Result:** No SQL errors related to invalid default values for columns of type `text` or in case of MySQL also `mediumtext`.

**Step 7:** Using a tool like e.g. PhpMyAdmin or PhpPgAdmin, verify in Joomla database tables that none of the table columns of SQL type `text` or in case of MySQL also `mediumtext` has a default value.

**Result:** No database table columns of SQL type `text` or in case of MySQL also `mediumtext` with a default value.

#### Test 2 - Update SQL scripts

**Step 1:** On a clean, current staging branch or nightly build or a 3.9.15 without the patch of this PR applied, or on an older Joomla version, make a new installation.

**Step 2:** Apply the patch of this PR.

**Step 3:** Login to backend.

**Step 4:** Go to `Extensions -> Manage -> Database`.

**Result:** Depends on the databasase type and the Joomla version.
- MySQL: On current staging or nightly, only "1 Problem: Database schema version not up to date" is shown. For older Joomla versions there might be some database errors shown similar to those mentioned for PostgreSQL. These errors then refer to wrong database columns of type `text` or `mediumtext`.
- PostgreSQL: On current staging or nightly, 34 database problems are found, which all belong to update sql script `3.9.16-2020-02-15.sql`. The problems refer to wrong database columns of type `text`. For older Joomla versions there might be shown more database errors of this type.

**Step 5:** Use the "Fix" button.

**Result:** No database problems shown. Database schema version is `3.9.16-2020-02-15`.

**Step 6:** Follow the testing instructions of PR #17860 , i.e.:
- Go to administrator > Modules and set it to Administrator.
- Try to create a new Administrator Menu (for example), so that there is no editor tool in the creation window (because that's where the content comes from).
- Save this module.

**Result:** No SQL error, module is saved.

**Step 7:** Verify that following works as before and there are no SQL errors shown which are related to invalid default values for columns of type `text` or in case of MySQL also `mediumtext`:
- Creating and editing categories
- Creating and editing fields and field groups
- Smart Search
- Action logs
- Content history

**Result:** No SQL errors related to invalid default values for columns of type `text` or in case of MySQL also `mediumtext`.

**Step 8:** Using a tool like e.g. PhpMyAdmin or PhpPgAdmin, verify in Joomla database tables that none of the table columns of SQL type `text` or in case of MySQL also `mediumtext` has a default value.

**Result:** No database table columns of SQL type `text` or in case of MySQL also `mediumtext` with a default value.

#### Test 3 - Update to patched nightly build of today plus this PR applied

**Step 1:** On a Joomla 3.9.15 or any other 3.9.x but not current staging or nightly, go to the Joomla Update Component, change update channel to "Custom URL" and enter following custom URL:
[https://test5.richard-fath.de/test-pr-27937_list.xml](https://test5.richard-fath.de/test-pr-27937_list.xml).
Or download the patched update comtainer from here
[https://test5.richard-fath.de/Joomla_3.9.16-dev-Development-Update_Package_2020-02-15_pr-27937.zip](https://test5.richard-fath.de/Joomla_3.9.16-dev-Development-Update_Package_2020-02-15_pr-27937.zip)
and use the "Upload & Update" functionality of the Joomla Update component, then you can also update a current staging or nightly.

**Step 2:** Start the update.

**Result:** Update finishs with success.

**Step 3.** Go to `Extensions -> Manage -> Database`.

**Result:** No database problems shown. Database schema version is `3.9.16-2020-02-15`.

**Step 4:** Follow the testing instructions of PR #17860 , i.e.:
- Go to administrator > Modules and set it to Administrator.
- Try to create a new Administrator Menu (for example), so that there is no editor tool in the creation window (because that's where the content comes from).
- Save this module.

**Result:** No SQL error, module is saved.

**Step 5:** Verify that following works as before and there are no SQL errors shown which are related to invalid default values for columns of type `text` or in case of MySQL also `mediumtext`:
- Creating and editing categories
- Creating and editing fields and field groups
- Smart Search
- Action logs
- Content history

**Result:** No SQL errors related to invalid default values for columns of type `text` or in case of MySQL also `mediumtext`.

**Step 6:** Using a tool like e.g. PhpMyAdmin or PhpPgAdmin, verify in Joomla database tables that none of the table columns of SQL type `text` or in case of MySQL also `mediumtext` has a default value.

**Result:** No database table columns of SQL type `text` or in case of MySQL also `mediumtext` with a default value.

### Expected result

- No database table columns of SQL type `text` or in case of MySQL also `mediumtext` has a default value.
- No SQL error when creating an admin module as descibed in the testing instructions of PR #17860 .
- No SQL errors related to invalid default values for columns of type `text` or in case of MySQL also `mediumtext` in other cases.

### Actual result

- There are database table columns of SQL type `text` or in case of MySQL also `mediumtext` which have a default value.
- You might get SQL error when creating an admin module as descibed in the testing instructions of PR #17860 .
- Depending on your Joomla version you might such SQL errors in other cases, too.

### Documentation Changes Required

None.
